jib-core/src/main/java/com/google/cloud/tools/jib/frontend/MainClassFinder.java

```diff
@@ -109,7 +112,16 @@ public static String resolveMainClass(
 
         try {
           // Adds each file in the classes output directory to the classes files list.
-          List<String> mainClasses = findMainClasses(rootDirectory);
+          HashSet<Path> visitedRoots = new HashSet<>();
```

Comment:
Generally, in this codebase, we try to use the highest level type possible to limit exposed APIs as much as possible.